types/src/extended_data_square.rs

```diff
@@ -0,0 +1,150 @@
+use std::result::Result as StdResult;
+
+use nmt_rs::NamespaceMerkleHasher;
+
+use crate::namespaced_data::{NamespacedData, NamespacedDataId};
+use crate::nmt::{Namespace, NamespacedSha2Hasher, Nmt};
+use crate::{DataAvailabilityHeader, Share};
+use crate::{Error, Result};
+
+/// Represents either Column or Row of the Data Square.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+#[repr(u8)]
+pub enum AxisType {
+    Row = 0,
+    Col,
+}
+
+impl TryFrom<u8> for AxisType {
+    type Error = Error;
+
+    fn try_from(value: u8) -> StdResult<Self, Self::Error> {
+        match value {
+            0 => Ok(AxisType::Row),
+            1 => Ok(AxisType::Col),
+            n => Err(Error::InvalidAxis(n.into())),
+        }
+    }
+}
+
+/// Represents ExtendedDataSquare with deserialized Shares
+pub struct ExtendedDataSquare {
+    shares: Vec<Share>,
+    square_len: usize,
+}
+
+impl ExtendedDataSquare {
+    /// Create a new EDS out of the provided shares. Returns error if share number doesn't
+    /// create a square
+    pub fn new(shares: Vec<Share>) -> Result<Self> {
+        let square_len = f64::sqrt(shares.len() as f64) as usize;
+        if square_len * square_len != shares.len() {
+            return Err(Error::EdsInvalidDimentions);
+        }
+
+        Ok(Self { shares, square_len })
+    }
+
+    /// Return row with index
+    pub fn row(&self, index: usize) -> Result<Vec<Share>> {
+        self.shares
+            .get(index * self.square_len..(index + 1) * self.square_len)
+            .ok_or(Error::EdsIndexOutOfRange(index))
+            .map(|v| v.to_vec())
+    }
+
+    /// Return colum with index
+    pub fn column(&self, mut index: usize) -> Result<Vec<Share>> {
+        let mut r = Vec::with_capacity(self.square_len);
+        while index < self.shares.len() {
+            r.push(
+                self.shares
+                    .get(index)
+                    .ok_or(Error::EdsIndexOutOfRange(index))?
+                    .clone(),
+            );
+            index += self.square_len;
+        }
+        Ok(r)
+    }
+
+    /// Return column or row with the provided index
+    pub fn axis(&self, axis: AxisType, index: usize) -> Result<Vec<Share>> {
+        match axis {
+            AxisType::Col => self.column(index),
+            AxisType::Row => self.row(index),
+        }
+    }
+
+    /// Get EDS square length
+    pub fn square_len(&self) -> usize {
+        self.square_len
+    }
+
+    /// Return all the shares that belong to the provided namespace in the EDS.
+    /// Results are returned as a list of rows of shares with the inclusion proof
+    pub fn get_namespaced_data(
```

Comment:
could we add a test for that?

Reply:
I've added a test (now in `rsmt2d.rs`). I don't think we have a code generating entire DAH from EDS, so I've mocked the DAH manually as this is getting scope creepy already.

Reply:
ooh I was thinking about fetching extended header and then calling get eds for it and save it in test data, and here just use

Reply:
so you can eg. modify the celestia-rpc get_eds test and submit sample a bit bigger blob, like 1000b, then submit it and fetch the data for that block

Reply:
or you can submit many blobs with different namespaces

Reply:
Good idea, I've captured EDS and DAH using tests used that for testing instead 👍